pkg/operator/ceph/webhook.go

```diff
@@ -18,48 +18,58 @@ package operator
 
 import (
 	"context"
+	"fmt"
+	"io/ioutil"
 	"os"
 
 	"github.com/pkg/errors"
 	"github.com/rook/rook/pkg/clusterd"
-	"github.com/rook/rook/pkg/operator/ceph/controller"
-	"github.com/rook/rook/pkg/operator/ceph/csi"
 	"github.com/rook/rook/pkg/operator/k8sutil"
-	v1 "k8s.io/api/apps/v1"
 	corev1 "k8s.io/api/core/v1"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/intstr"
-	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
 )
 
 const (
-	appName                                  = "rook-ceph-admission-controller"
-	secretVolumeName                         = "webhook-certificates" // #nosec G101 This is just a var name, not a real secret
-	serviceAccountName                       = "rook-ceph-admission-controller"
-	portName                                 = "webhook-api"
-	servicePort                        int32 = 443
-	serverPort                         int32 = 8079
-	tlsDir                                   = "/etc/webhook"
-	admissionControllerTolerationsEnv        = "ADMISSION_CONTROLLER_TOLERATIONS"
-	admissionControllerNodeAffinityEnv       = "ADMISSION_CONTROLLER_NODE_AFFINITY"
+	appName       = "rook-ceph-admission-controller"
+	tlsPort int32 = 443
 )
 
 var (
 	namespace = os.Getenv(k8sutil.PodNamespaceEnvVar)
 )
 
 func isSecretPresent(ctx context.Context, context *clusterd.Context) (bool, error) {
-	logger.Infof("looking for secret %q", appName)
-	_, err := context.Clientset.CoreV1().Secrets(namespace).Get(ctx, appName, metav1.GetOptions{})
+	logger.Infof("looking for admission webhook secret %q", appName)
+	s, err := context.Clientset.CoreV1().Secrets(namespace).Get(ctx, appName, metav1.GetOptions{})
 	if err != nil {
 		// If secret is not found. All good ! Proceed with rook without admission controllers
 		if apierrors.IsNotFound(err) {
-			logger.Infof("secret %q not found. proceeding without the admission controller", appName)
+			logger.Infof("admission webhook secret %q not found. proceeding without the admission controller", appName)
 			return false, nil
 		}
 		return false, err
 	}
+
+	// Search for any previous admission controller deployment and if so removing it
+	logger.Debug("searching for old admission controller deployment")
+	removeOldAdmissionControllerDeployment(ctx, context)
+
+	logger.Infof("admission webhook secret %q found", appName)
+	err = os.Mkdir(certDir, 0400)
```

Comment:
Instead of (or maybe in addition to) mkdir, let's define the directory in the operator deployment (and helm chart) as a volume. Otherwise, the directory may fail to create on more restrictive environments such as openshift.

Reply:
Done in-lieu of the Mkdir.